### PR TITLE
Switch to IndexedDB

### LINKED ISSUE #256: Switch to IndexedDB
IndexedDB allows storing raw objects that don't need to be serialized to a string. So we could store the splits as raw byte buffers instead, which we need in livesplit-core anyway, rather than JS strings. 
Additionally it seems like browsers allow it to be larger than normal localStorage, so this would also help with #70 

It seems like the API for IndexedDB is quite obscure though, so even the documentation for it recommends using something like [localForage](https://localforage.github.io/localForage/) instead which wraps it nicely. Though it brings a lot of baggage of trying to work on older browsers, but we basically only target browsers that at least support WebAssembly. So those definitely support IndexedDB afaict, so shipping any kind of polyfills is just bloating our download size.

### PR DESCRIPTION
**Pull request by wooferzfg**
Fix #256

I'm opening this as a draft PR because I'm not sure how to approach storing the data as raw bytes or implementing compression for the splits files.

### CODE REVIEW

**Comment by DarkRTA** [comment]
This will also close #70 as it'll no longer apply after merging this.

File: src/ui/LiveSplit.tsx [hunk 1]
```diff
@@ -39,6 +40,13 @@ type Menu =
     { kind: MenuKind.SettingsEditor, config: HotkeyConfig } |
     { kind: MenuKind.About };
 
+export interface Props {
+    splits: string,
+    layout: any,
+    settings: any,
```
Comment (CryZe) [inline src/ui/LiveSplit.tsx]: These probably shouldn't be any.

File: src/ui/LiveSplit.tsx [hunk 2]
```diff
@@ -39,6 +40,13 @@ type Menu =
     { kind: MenuKind.SettingsEditor, config: HotkeyConfig } |
     { kind: MenuKind.About };
 
+export interface Props {
+    splits: string,
```
Comment (CryZe) [inline src/ui/LiveSplit.tsx]: The splits should be ArrayBuffer / Uint8Array.

File: src/ui/LiveSplit.tsx [hunk 3]
```diff
@@ -1,4 +1,5 @@
 import * as React from "react";
+import localforage from "localforage";
```
Comment (CryZe) [inline src/ui/LiveSplit.tsx]: I'm not sure if we want localforage or not. If we manually talk to IndexedDB we could have tables for splits and for layouts, which would help us for the run and layout selection view.

**Comment by CryZe** [comment]
Also considering I'm seeing a lot of people use LSO by now, I believe we need to have backwards compatibility with localStorage, so people don't lose their splits.

File: src/ui/LiveSplit.tsx (hunk 1 above)
Comment (CryZe) [reply src/ui/LiveSplit.tsx]: In fact I believe that we even should "ban any". They somewhat recently introduced `unknown` as a [top type](https://en.wikipedia.org/wiki/Top_type) that doesn't imply dynamic typing.

**Comment by CryZe** [comment]
I just implemented the byte buffer storing.

**Comment by wooferzfg** [comment]
We still need to implement backwards compatibility for localStorage so that splits, layouts, and settings aren't lost once this is merged.

**Comment by wooferzfg** [comment]
Let's make sure that the livesplit-core commits are on master before merging this.

File: src/storage/index.tsx [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+                db.createObjectStore("splits", {
+                    autoIncrement: true,
+                });
+                db.createObjectStore("layouts", {
+                    autoIncrement: true,
+                });
+                db.createObjectStore("settings", {
+                    autoIncrement: true,
+                });
+            },
```
Comment (CryZe) [inline src/storage/index.tsx]: I probably should've told you, but IndexedDB has a notion of migrating between different versions. The migration from no IndexedDB to version 1 is also a migration. This is being done here. And this is the only place we ever need to migrate from localStorage to IndexedDB. So all the migration code should be in here.

Comment (wooferzfg) [reply src/storage/index.tsx]: So would we just load everything from localStorage inside `upgrade` and immediately add them to the database?

Comment (CryZe) [reply src/storage/index.tsx]: Yes, I'll push a fix for it in a few minutes.

**Pull request merged**